packages/ra-core/src/controller/useListController.ts

```diff
@@ -158,14 +160,32 @@ const useListController = (props: ListProps): ListControllerProps => {
                     'warning'
                 ),
         },
-        (state: ReduxState) =>
-            state.admin.resources[resource]
-                ? state.admin.resources[resource].list.ids
-                : null,
-        (state: ReduxState) =>
-            state.admin.resources[resource]
-                ? state.admin.resources[resource].list.total
-                : null
+        // data selector
+        (state: ReduxState): Identifier[] => {
+            const resourceState = state.admin.resources[resource];
+            // if the resource isn't initialized, return null
+            if (!resourceState) return null;
+            const idsForQuery =
+                resourceState.list.idsForQuery[requestSignature];
```

Comment:
Would'nt it be better to have a dataForQuery holding both ids and total? This way the requestSignature would be present only once in the state.

Reply:
It would be better indeed, but I don't know how to implement it properly. 

Let's say we add a `cachedRequests` key to the `list` reducer:

```js
// instead of
const listState = {
    ids: [13, 12, 11, 10, 9, 8, 7, 6, 4, 2],
    loadedOnce: true,
    params: {
        sort: null,
        order: null,
        page: 1,
        perPage: null,
        filter: {}
    },
    selectedIds: [],
    total: 13,
    idsForQuery: {
        '{"pagination":{"page":1,"perPage":10},"sort":{"field":"published_at","order":"DESC"},"filter":{}}': [13, 12, 11, 10, 9, 8, 7, 6, 4, 2],
    },
    totalForQuery: {
        '{"pagination":{"page":1,"perPage":10},"sort":{"field":"published_at","order":"DESC"},"filter":{}}': 13
    },
    validity: {
        '{"pagination":{"page":1,"perPage":10},"sort":{"field":"published_at","order":"DESC"},"filter":{}}': '2020-02-24T05:28:13.220Z'
    }
};

// group cache-related state as
const listState = {
    ids: [13, 12, 11, 10, 9, 8, 7, 6, 4, 2],
    loadedOnce: true,
    params: {
        sort: null,
        order: null,
        page: 1,
        perPage: null,
        filter: {}
    },
    selectedIds: [],
    total: 13,
    cachedRequests: {
        '{"pagination":{"page":1,"perPage":10},"sort":{"field":"published_at","order":"DESC"},"filter":{}}': {
            ids: [13, 12, 11, 10, 9, 8, 7, 6, 4, 2],
            total: 13,
            validity: '2020-02-24T05:28:13.220Z',
        }
    }
};
```

The problem is: How do I compose the `ids`, `total`, and `validity` reducers into a variable key of the `cachedRequests` reducer? We've done that for the `resources` reducer and it's hardly readable.

I'm still scratching my head about this one.

Reply:
I've found a way, will commit soon